Mathlib/RingTheory/Radical.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2024 Jineon Back and Seewoo Lee. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jineon Baek, Seewoo Lee
+-/
+import Mathlib.Algebra.Ring.Regular
+import Mathlib.RingTheory.UniqueFactorizationDomain
+
+/-!
+# Radical of an element of a unique factorization normalization monoid
+
+This file defines a radical of an element `a` of a unique factorization normalization
+monoid, which is defined as a product of normalized prime factors of `a` without duplication.
+This is different from the radical of an ideal.
+
+## Main declarations
+
+- `radical`: The radical of an element `a` in a unique factorization monoid is the product of
+  its prime factors.
+- `radical_eq_of_associated`: If `a` and `b` are associates, i.e. `a * u = b` for some unit `u`,
+  then `radical a = radical b`.
+- `radical_unit_hMul`: Multiplying unit does not change the radical.
+- `radical_dvd_self`: `radical a` divides `a`.
+- `radical_pow`: `radical (a ^ n) = radical a` for any `n ≥ 1`
+- `radical_prime`: Radical of a prime element is equal to its normalization
+
+## TODO
+
+- Make a comparison with `Ideal.radical`. Especially, for principal ideal,
+  `Ideal.radical (Ideal.span {a}) = Ideal.span {radical a}`.
+- Prove `radical (radical a) = radical a`.
+-/
+
+noncomputable section
+
+open scoped Classical
+
+open UniqueFactorizationMonoid
+
+variable {k : Type _} [Field k]
+-- `CancelCommMonoidWithZero` is required by `UniqueFactorizationMonoid`
+variable {α : Type _} [CancelCommMonoidWithZero α] [NormalizationMonoid α]
+  [UniqueFactorizationMonoid α]
+variable {R : Type _} [CommRing R] [IsDomain R] [NormalizationMonoid R] [UniqueFactorizationMonoid R]
```

Comment:
Rather than declaring these at the front, I'd prefer if the contents are split into sections so that users will not have to distinguish the typeclass assumptions by looking at the type name and guessing. 
Also please use `Type*` instead of `Type _`.